File: my-folder/0908-middle-of-the-linked-list/solution.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function middleNode(head: ListNode | null): ListNode | null {
    const length = noOfNodes(head, 0);
    const mid = Math.floor(length / 2) + 1;

    return (length === 0)? null : findNode(head, 1, mid); 
};

function findNode(head: ListNode, idx: number, target: number): ListNode {
    if (idx === target) return head;

    return findNode(head.next, idx + 1, target);
}

function noOfNodes(head: ListNode | null, size: number): number {
    if (head === null) {
        return size;
    } else {
        return noOfNodes(head.next, size + 1)
    }
}

File: my-folder/0908-middle-of-the-linked-list/solution.py
from typing import Tuple
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        (middle_node, length) = self.get_middle(head, 0, 0)
        return middle_node
        
    def get_middle(self, head: Optional[ListNode], idx: int, length: int) -> Tuple[Optional[ListNode], int]:
        if not head:
            return (None, length)
        
        (new_head, length) = self.get_middle(head.next, idx + 1, length + 1)

        if length % 2 == 0 and idx == length / 2:
            return (head, length)
        elif length % 2 != 0 and idx == length // 2:
            return (head, length)
        else:
            return  (new_head, length)
        


